node/core/dispute-coordinator/src/participation/queues/mod.rs

```diff
@@ -251,7 +262,16 @@ impl Queues {
 			if self.best_effort.len() >= BEST_EFFORT_QUEUE_SIZE {
 				return Err(QueueError::BestEffortFull)
 			}
-			self.best_effort.insert(comparator, req);
+			// Keeping old request if any. This prevents request timers
+			// from resetting on each new request at the same priority
+			// level for the same candidate.
+			if self.best_effort.contains_key(&comparator) {
+				if let Some(timer) = req.request_timer {
+					timer.stop_and_discard();
+				}
+			} else {
+				self.best_effort.insert(comparator, req);
```

Comment:
You can avoid the double lookup via [`entry`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.entry).

Reply:
Maybe I'm missing something obvious, but here's the problem I see there. The intuitive pattern is something like:

`self.priority.entry(comparator).or_insert(req);`

But this pattern drops `req` without calling `stop_and_discard()` on its timer if there was an older request. This sends bad data to our metrics. I don't see a way to call `stop_and_discard()` without using a double lookup. 🤷‍♂️

Reply:
Can't you use [or_insert_with](https://doc.rust-lang.org/std/collections/btree_map/enum.Entry.html#method.or_insert_with) to execute `timer.stop_and_discard()` first and then insert `req`?

Reply:
you can also always pattern match on the `Entry` constructors and do whatever you need to do. See [here](https://github.com/paritytech/polkadot/blob/dfa8c5ae321e0523f88fdc43afc6d743812ee49e/statement-table/src/generic.rs#L391) for an example.

Reply:
Your solution worked @eskimor Much appreciated!